Sources/NIO/CircularBuffer.swift

```diff
@@ -234,3 +182,131 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
+    ///
+    /// - Parameter newElements:
+    /// The new elements to add to the collection.
+    ///
+    /// *O(n)* where _n_ is the length of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange<C>(_ subrange: Range<Index>, with newElements: C) where C : Collection, E == C.Element {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex, "Subrange out of bounds")
+
+        if subrange.count == newElements.count {
+            // Can't just zip(subrange, newElements) because the compiler complains about:
+            // «argument type 'Range<Int>' does not conform to expected type 'Sequence'»
+            // with Swift version 4.1.2 (swiftlang-902.0.54 clang-902.0.39.2)
+            for (index, element) in zip(subrange.lowerBound..<subrange.upperBound, newElements) {
+                self.buffer[self.bufferIndex(ofIndex: index)] = element
+            }
+        } else if subrange.count == self.count && newElements.isEmpty {
+            self.removeSubrange(subrange)
+        } else {
+            var newBuffer: ContiguousArray<E?> = []
+            let capacityDelta = (Int(newElements.count) - subrange.count)
+            let newCapacity = Int(UInt32(self.buffer.count + capacityDelta).nextPowerOf2())
+            newBuffer.reserveCapacity(newCapacity)
+
+            // This mapping is required due to an inconsistent ability to append sequences of non-optional
+            // to optional sequences.
+            // https://bugs.swift.org/browse/SR-7921
+            newBuffer.append(contentsOf: self[0..<subrange.lowerBound].lazy.map { $0 })
+            newBuffer.append(contentsOf: newElements.lazy.map { $0 })
+            newBuffer.append(contentsOf: self[subrange.upperBound..<self.endIndex].lazy.map { $0 })
+
+            self.tailIdx = newBuffer.count
+            let repetitionCount = newCapacity - newBuffer.count
+            if repetitionCount > 0 {
+                newBuffer.append(contentsOf: repeatElement(nil, count: repetitionCount))
+            }
+            self.headIdx = 0
+            self.buffer = newBuffer
+        }
+    }
+
+    /// Removes the elements in the specified subrange from the circular buffer.
+    ///
+    /// - Parameter bounds: The range of the circular buffer to be removed. The bounds of the range must be valid indices of the collection.
+    public mutating func removeSubrange(_ bounds: Range<Int>) {
+        precondition(bounds.upperBound >= self.startIndex && bounds.upperBound <= self.endIndex, "Invalid bounds.")
+        if bounds.count == 1 {
+            _ = remove(at: bounds.lowerBound)
+        } else if bounds.count == self.count {
+            self = .init(initialRingCapacity: self.buffer.count)
+        } else {
```

Comment:
nit: should we use a switch here ?

Reply:
I think a `switch` would make this a lot more readable.

I'll change it during the day